### ISSUE TITLE
CosFace Layer Accuracy

### ISSUE THREAD

**Issue by abnera** [MuggleWang/CosFace_pytorch#4]
I understand the CosFace layer is applied to the feature vector of a given model. Then the output of `cosface` is then passed to the criterion (`cross-entropy`) in order to the loss. 

However, how to calculate accuracy is I have a validation dataset that is a classification problem. I know that for LFW we use the features and cosine similarity but I want regular classification accuracy assuming both the train and val have the same classes. I tried using the output of `cosface` layer as the logits for accuracy metric since they have same shape as target/labels but that is leading to 0% accuracy even if the loss seems to be training properly and same problem with just a linear + cross-entropy give good results. 

Basically, I want to use the output of `cosface` layer to calculate train and validation accuracy.

Also how to select `s` parameter if you have any thoughts.

**Comment by MuggleWang** [comment]
You want to calculate classification accuracy, use variable `cosine`  . The biggest one in `cosine` is predicted label. You can think about the physical meaning of each variable.
We generally think that face recognition is an open-set problem. The accuracy of classification is of little significance. You can check out the test methods on the public challenge set.
The reason we multiply the parameter s is because we want the output of the cosface layer to be the same as the linear layer, the value range is [-∞, +∞ ]. Thus, the value of the softmax activation layer is stable(0, 1). In experience, 30 and 64 are reasonable. Further, you can use the linear layer training model to observe the length of linear layer output of the trained model. s should be numerically similar to it

**Comment by dattx0** [comment]
@MuggleWang can you explain why using cosine in softmax for validation not the output of MCP?

**Comment by MuggleWang** [comment]
Variable `cosine`  is inner product of feature `input` and class weight `self.weight`.  Assuming we get only one feature, which means dim of `input` is `1*512`. Assuming number of classes is `c`, then dim of `self.weight` is 'c*512. Dim of `cosine` is '1*c'. Each value in `cosine` means cosine similarity of the feature and `c` weights.
Output of this layer is cosine minus m in label of the feature. It reduces the confidence rate at the feature label.

**Comment by abnera** [comment]
Great explanation @MuggleWang this makes perfect sense. I figured some of these things after posting the question but now I can be certain. 

One final clarification: 
`Further, you can use the linear layer training model to observe the length of linear layer output of the trained model. s should be numerically similar to it.` 
Can you elaborate on this by length of linear layer you mean the shape? wouldn't the shape always be the same (batch_size, num_classes) that's the shape of output. 
I was thinking of making `s` value such that when multiply the values of output as similar to those of if only using a linear layer instead of cosface layer.

**Comment by MuggleWang** [comment]
@abnera Sorry, this is a mistake.
Actually, m should be similar to `||W_j|| ||x||`. Where `W_j` is j-th weight of linear layer and `x` is a feature vector. You can observe the changes in the training process which will be very interesting.
At first, `the length of linear layer output` means `L2 norm of the vector`. Then, I realized that it is still inaccurate.
